src/sources/util/http_scrape.rs

```diff
@@ -0,0 +1,199 @@
+    Error, SourceSender,
+};
+use vector_common::shutdown::ShutdownSignal;
+use vector_core::{config::proxy::ProxyConfig, event::Event, ByteSizeOf};
+
+/// Contains the inputs generic to any http scrape.
+pub(crate) struct GenericHttpScrapeInputs {
+    /// Array of URLs to scrape
+    pub urls: Vec<Uri>,
+    /// Interval to scrape on in seconds
+    pub interval_secs: u64,
+    /// Map of Header+Value to apply to HTTP request
+    pub headers: HashMap<String, Vec<String>>,
+    /// Content type of the HTTP request, determined by the source
+    pub content_type: String,
+    pub auth: Option<Auth>,
+    pub tls: TlsSettings,
+    pub proxy: ProxyConfig,
+    pub shutdown: ShutdownSignal,
+}
+
+/// The default interval to scrape the http endpoint if none is configured.
+pub(crate) const fn default_scrape_interval_secs() -> u64 {
+    15
+}
+
+/// Methods that allow context-specific behavior during the scraping procedure.
+pub(crate) trait HttpScraper {
+    /// (Optional) Called before the HTTP request is made, allows building context.
+    fn build(&mut self, _url: &Uri) {}
+
+    /// Called after the HTTP request succeeds and returns the decoded/parsed Event array.
+    fn on_response(&mut self, url: &Uri, header: &Parts, body: &Bytes) -> Option<Vec<Event>>;
+
+    /// (Optional) Called if the HTTP response is not 200 ('OK').
+    fn on_http_response_error(&self, _uri: &Uri, _header: &Parts) {}
+}
+
+/// Builds a url for the HTTP requests.
+pub(crate) fn get_url(uri: &Uri, query: &HashMap<String, Vec<String>>) -> Uri {
+    let mut serializer = url::form_urlencoded::Serializer::new(String::new());
+    if let Some(query) = uri.query() {
+        serializer.extend_pairs(url::form_urlencoded::parse(query.as_bytes()));
+    };
+    for (k, l) in query {
+        for v in l {
+            serializer.append_pair(k, v);
+        }
+    }
+    let mut builder = Uri::builder();
+    if let Some(scheme) = uri.scheme() {
+        builder = builder.scheme(scheme.clone());
+    };
+    if let Some(authority) = uri.authority() {
+        builder = builder.authority(authority.clone());
+    };
+    builder = builder.path_and_query(match serializer.finish() {
+        query if !query.is_empty() => format!("{}?{}", uri.path(), query),
+        _ => uri.path().to_string(),
+    });
+    builder
+        .build()
+        .expect("Failed to build URI from parsed arguments")
+}
+
+/// Scrapes one or more urls at an interval.
+///   - The HTTP request is built per the options in provided generic inputs.
+///   - The HTTP response is decoded/parsed into events by the specific context.
+///   - The events are then sent to the output stream.
+pub(crate) async fn http_scrape<H: HttpScraper + std::marker::Send + Clone>(
+    inputs: GenericHttpScrapeInputs,
+    context: H,
+    mut out: SourceSender,
+) -> Result<(), ()> {
+    let mut stream = IntervalStream::new(tokio::time::interval(Duration::from_secs(
+        inputs.interval_secs,
+    )))
```

Comment:
I'm not familiar with the current approach to scraping. Could this cause multiple requests to be outstanding simultaneously if a slow server causes a timeout longer than `interval_secs`?

This, incidentally, is one reason to avoid mutable state in per-request context data.

Reply:
That's a good question. I don't have as good of an idea of the answer as the original author might have. But I would suspect that a slow server would probably cause problems like that. I'm investing some time in trying to produce those conditions (slow server) , but not having luck so far.

Reply:
Since I'm out of runway- would you be opposed to me creating a follow-up issue to investigate this?

I don't currently know enough about how these streams are handled to be able to talk intelligently about it.
Could also try to set up a mock server that would cause a timeout to try to trigger the behavior in question.

Reply:
> I'm investing some time in trying to produce those conditions (slow server) , but not having luck so far.

> Could also try to set up a mock server that would cause a timeout to try to trigger the behavior in question.

Hi @neuronull , with Python and `aiohttp` you can easily produce such a mock server. See below. Hope it helps!

```python
import asyncio
import random
from aiohttp import web

async def handler(request):
    delay = random.random() * 2 + 1
    await asyncio.sleep(delay)
    return web.Response(text=f"request took {delay}s to complete\n")

app = web.Application()
app.add_routes([web.get('/', handler)])  # can be web.post(...) etc
web.run_app(app)
```
```
$ python server.py
======== Running on http://0.0.0.0:8080 ========
(Press CTRL+C to quit)
```
```
$ time curl localhost:8080/
request took 1.3066107637425408s to complete

real    0m1.319s
user    0m0.009s
sys     0m0.000s

$ time curl localhost:8080/
request took 2.666990961438348s to complete

real    0m2.680s
user    0m0.008s
sys     0m0.000s
```

Reply:
Hey @hhromic , thanks for that, worked great!
What I did is I set up the server to delay 10s on first request, and 1s on all subsequent requests.
Set the scrape interval to 1s.
It seems that `aiohttp` should be able to handle multiple requests in parallel due to its async support.
The result however was that the first request took 10s to complete, and all subsequent requests 1s after. It appears you may be on to something @bruceg ... Though given that this logic was part of the original `prometheus_scrape` (and it appears, other scrape-ers in the project) , what are your thoughts on creating a follow-up issue for that?

Reply:
I think that is good enough at this point, yes.

Reply:
:+1: I opened https://github.com/vectordotdev/vector/issues/14087

Reply:
> It seems that `aiohttp` should be able to handle multiple requests in parallel due to its async support.

Yes, almost :)
Indeed `aiohttp` will handle thousands of requests _concurrently_, but not in parallel.

> The result however was that the first request took 10s to complete, and all subsequent requests 1s after.

The actual behaviour depends on how you implemented the logic. I just took the exercise for fun, see below.
As you can see, the total execution for the three requests is 10s, with the first 2 completed in the first second.

```python
import asyncio
import random
from aiohttp import web

async def handler(request):
    data = request.config_dict["data"]
    data["req_num"] += 1
    req_num = data["req_num"]  # store locally as it can change during awaits
    delay = 10 if req_num == 1 else 1
    await asyncio.sleep(delay)
    return web.Response(text=f"req {req_num} = {delay}s\n")

app = web.Application()
app.add_routes([web.get("/", handler)])  # can be web.post(...) etc
app["data"] = {"req_num": 0}
web.run_app(app)
```
```
$ time bash -c 'for _ in 1 2 3; do time curl -s localhost:8080 & done; wait'
req 2 = 1s
req 3 = 1s

real    0m1.022s
user    0m0.019s
sys     0m0.000s

real    0m1.023s
user    0m0.018s
sys     0m0.000s
req 1 = 10s

real    0m10.028s
user    0m0.009s
sys     0m0.009s

real    0m10.033s
user    0m0.056s
sys     0m0.009s
```

Reply:
Thanks for the follow-up!
I tested that out against the source and got an unexpected result, will document it in the new issue opened and we can dig deeper into it then :+1: